### DISCUSSION TITLE
Server Component calling server action getting Cookies can only be modified in a Server Action or Route Handler. error

### DISCUSSION THREAD

**Discussion by the-devbear** [vercel/next.js#60025] [Help] [answered]
### Summary

Hi, I am new to NextJS and I need some help.
I have this Server action: 
```typescript
'use server';

import { cookies } from 'next/headers';

export async function getData() {
  const cookieStore = cookies();

  cookieStore.set('test', 'other value');
  const cookie = cookieStore.get('test');
  console.log(cookie);

  return {
    hello: 'world',
    cookie,
  };
}
```
and this is my Component:
```typescript
import { getData } from '@/lib/actions/data-test';

export default async function TestPage() {
  const data = await getData();

  return <div>{JSON.stringify(data)}</div>;
}
```
But I am getting this error: `Error: Cookies can only be modified in a Server Action or Route Handler.`
But I don't know why. 
This is just an example for simplicity but I need to set cookies, because I am using the Spotify API and want to if need set the refresh and access token in the cookies. 
What did I do wrong?

Thanks in advance 👍 

### Additional information

```js
Next info:

Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.1.0: Mon Oct  9 21:28:12 PDT 2023; root:xnu-10002.41.9~6/RELEASE_ARM64_T8103
Binaries:
  Node: 18.17.1
  npm: 9.6.7
  Yarn: 1.22.19
  pnpm: N/A
Relevant Packages:
  next: 14.0.4
  eslint-config-next: 14.0.4
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.3.3
Next.js Config:
  output: N/A


warn  - Latest canary version not detected, detected: "14.0.4", newest: "14.0.5-canary.30".
        Please try the latest canary version (`npm install next@canary`) to confirm the issue still exists before creating a new issue.
        Read more - https://nextjs.org/docs/messages/opening-an-issue
```
```


### Example

_No response_

**Comment by ananuness** [comment, marked as answer]
This means is that you cannot support this behavior with just Server Components, you need to use either:

- An API call that sets the cookie
- Set the cookie in JS in the browser
- Use Server Actions
- Use middleware

You have those 4 ways to archive it without problems, you can see more about it in this [article](https://www.propelauth.com/post/cookies-in-next-js). 

Also, Server actions are used to handle form submissions and mutations, but it's not limited to it, you can invoke it in event handlers, like `useEffect` for example. See more about [Server Actions and its behavior here](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations?ref=propelauth.com#behavior).

Finally, as you can see you're just calling it directly in your Server component, that's why you got the error. Now, taking into account what was said, to make your work easier you can call `getData()` inside an useEffect or make `getData` as a Route handler.

Hope it helps! Don't hesitate to ask!

**Reply by the-devbear** [reply]
Thanks, that helps a lot. Small follow up question. 
Should I use server actions to get data or only to do an "action" like creating/updating etc.
Again thanks for the response 👍

**Reply by ananuness** [reply]
You're welcome!! And about server actions use cases, it's more common to use to execute actions like you mentioned (create, update etc) because of its asynchronous behaviour and server-side execution!

But there are more interesting uses that I think you would like to know! [Here](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations) you can see more examples. As for get data, the native fetch will do the job pretty well!

**Reply by seydanurhatipoglu** [reply]
Hey @ananuness, there is a point I don't understand about your answer. I am trying implement a custom i18n functionality in my application. My logic is to set a cookie "NEXT_LOCALE" when application is run initially. So, I am trying to invoke a server action in my root layout as below.

actions/i18n.ts
```
'use server'
import { cookies } from 'next/headers'

const defaultLocale: Locale = 'en'
const NEXT_LOCALE = 'NEXT_LOCALE'

export async function setLocaleCookieInitially() {
  if (!cookies().has(NEXT_LOCALE)) {
    changeLocale(defaultLocale)
  }
}
```

app/layout.tsx
```
export default async function RootLayout({ children }: BaseComponentProps) {
  await setLocaleCookieInitially()

  return (
    <html lang="en">
      <body className={poppins.className}>
        <Providers>
          {children}
        </Providers>
      </body>
    </html>
  )
}
```

this is throwing the same error as the op's question. As documentation clearly uses the same logic in actions in [here](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations?ref=propelauth.com#cookies), I don't understand why this is not working. Since this is in the root layout, I do not want to make it a client component and use useEffect hook and as I understand, server actions already make an api call under the hook. Thanks in advance.

**Reply by ananuness** [reply]
Hey @seydanurhatipoglu! Are you setting a cookie in your function `changeLocale`? Because cookies in server rendering are read-only.

And about Server Action, the thing is: `.set()` is available in Server actions, but used in an active part (it's commonly triggered by an action e.g sending form, pressing a button etc), it's not used automatically like just calling it directly in your Server component (e.g RootLayout). 

But if you're just verifying your cookie using `.has()`, then it's strange that you're getting this error 🧐

**Reply by seydanurhatipoglu** [reply]
Sorry I forgot to add `changeLocale` before, and yes, I am trying to set a cookie.

```
export async function changeLocale(locale: Locale) {
  cookies().set(NEXT_LOCALE, locale)
}
```

So, I guess setting it in middleware is my only option for my current requirements?

**Reply by ananuness** [reply]
You can use a route handler or middleware! I'm unsure which one should be the best for you! But you can start with middleware! And to help you out, you can check this [article](https://www.propelauth.com/post/cookies-in-next-js) that explains better the paths you can go to handle your cookies, if you need any additional help, just ask here!! 🙋‍♀️

**Reply by seydanurhatipoglu** [reply]
ok, thanks a lot!

**Reply by Juanryhn** [reply]
hello @ananuness currently I'm working in my project and have problem with cookies. My current flow is 
I have a login form and it will set a cookies after I got the response from external API. I hit an external API with server action and setting a cookies with it.
But now I want get the cookies inside route handler(this route handler will hit an external API) that I hit with server component. and it's got undefined. 
Do I have to change my server component to a client component ? So all of my component that need to get data from external API will be client component. If it so, do I will lost the benefit of SEO using nextjs ?

**Reply by Housi** [reply]
you can use a route handler but route handlers require absolute urls which require setting env vars for each of my environments xD 
and really after few hours of fighting with this I'm sure something won't be supported in my particular config anyway.

I can use middleware but then that wont work for navigation to already rendered routes

This is insane! so much problem with cookies... This should be fixed really, I used next for so many years but right now with v15 and app router I think I'm better off with some custom react setup. takes time, but maybe less than building apps with this complexity king, due to problems with most common web features

**Reply by tstiemerling** [reply]
I am following the example code given above, and calling an embedded server action function in page.tsx component, but I still get "Error: Cookies can only be modified in a Server Action or Route Handler.", so I am a little confused now. There must be some other subtleties to getting this working that are not documented.

**Reply by maiconsanson** [reply]
It's frustrating that we cannot set a simple cookie from a page (RSC) with server action and it is needed to use a `form`, a `client event`, or even a `route handler`.
